src/vm.rs

```diff
@@ -946,6 +962,52 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
         let page_tracker = self.vm_pages.page_tracker();
         page_tracker.rm_active_guest(self.vm_pages.page_owner_id());
     }
+
+    pub fn guest_set_shared_page_range(
+        &self,
+        guest_id: u64,
+        guest_addr: u64,
+        page_type: sbi::TsmPageType,
+        num_pages: u64,
+    ) -> sbi::Result<u64> {
+        if page_type != TsmPageType::Page4k || num_pages == 0 {
+            return Err(SbiError::InvalidParam);
+        }
+        let page_addr = self.guest_addr_from_raw(guest_addr)?;
+        let guest = self.guest_by_id(guest_id)?;
+        let guest_vm = guest.as_initializing_vm().ok_or(SbiError::InvalidParam)?;
+        guest_vm
+            .vm_pages
+            .set_shared_page_range(page_addr, num_pages)
+            .map_err(|_| SbiError::Failed)?;
+        Ok(num_pages)
+    }
+
+    pub fn guest_insert_shared_pages(
+        &self,
+        guest_id: u64,
+        page_addr: u64,
+        page_type: sbi::TsmPageType,
+        num_pages: u64,
+        guest_addr: u64,
+    ) -> sbi::Result<u64> {
+        if page_type != TsmPageType::Page4k || num_pages == 0 {
+            return Err(SbiError::InvalidParam);
+        }
+        let page_addr = self.guest_addr_from_raw(page_addr)?;
+        let guest = self.guest_by_id(guest_id)?;
+        if let Some(guest_vm) = guest.as_finalized_vm() {
```

Comment:
Do we want to the host to pre-fault shared pages?

Reply:
The original implementation did have it, but I removed it since I thought it probably makes sense to just say that shared pages are faulted-in on demand, but can always add it back.

Reply:
@abrestic-rivos What do you mean by pre-fault - do you mean pre-populate?

Reply:
Yup